helix-core/src/doc_formatter/test.rs

```diff
@@ -0,0 +1,232 @@
+use std::rc::Rc;
+
+use crate::doc_formatter::{DocumentFormatter, TextFormat};
+use crate::text_annotations::{InlineAnnotation, Overlay, TextAnnotations};
+
+impl TextFormat {
+    fn new_test(softwrap: bool) -> Self {
+        TextFormat {
+            soft_wrap: softwrap,
+            tab_width: 2,
+            max_wrap: 3,
+            max_indent_retain: 4,
+            wrap_indicator: ".".into(),
+            wrap_indicator_highlight: None,
+            // use a prime number to allow linging up too often with repear
+            viewport_width: 17,
+        }
+    }
+}
+
+impl<'t> DocumentFormatter<'t> {
+    fn collect_to_str(&mut self) -> String {
+        use std::fmt::Write;
+        let mut res = String::new();
+        let viewport_width = self.text_fmt.viewport_width;
+        let mut line = 0;
+
+        for (grapheme, pos) in self {
+            if pos.row != line {
+                line += 1;
+                assert_eq!(pos.row, line);
+                write!(res, "\n{}", ".".repeat(pos.col)).unwrap();
+                assert!(
+                    pos.col <= viewport_width as usize,
+                    "softwrapped failed {}<={viewport_width}",
+                    pos.col
+                );
+            }
+            write!(res, "{}", grapheme.grapheme).unwrap();
+        }
+
+        res
+    }
+}
+
+fn softwrap_text(text: &str, char_pos: usize) -> String {
```

Comment:
Right now `char_pos` is rather uninteresting because the `DocFormatter` laways start at the start of the line.
However once chunking is added in the future very long (4KB+) lines will be wrapped at locations which can be calculated in `O(1)` and then I will want to test that behavior too.

I actually already had this mostly implemented but I had an old helix instance running which overwrote all my work when I closed it :( I will open a PR for that once this is merged (just trying to keep this one from growing too much).

I removed the arguments for now in [692bbaa](https://github.com/helix-editor/helix/pull/5420/commits/692bbaa8e9b06a91d15d3ea01ce3b3945bfda87a) since it makes the tests more noisy. I might end up going with a different testing strategy for that particular feature anyway